transcendence-app/src/user/user.domain.ts

```diff
@@ -0,0 +1,10 @@
+export class User {
```

Comment:
NO! :smile: Let me explain

We should differentiate three different data models: domain model, entity model and dto. Each of them exists due to a specific reason.

- Domain model (this one, `User`) -> Represent the domain of our problem, despite now just being a dumb class holding data, as we grow our app it may hold business rules such as validation. Imagine we want to validate a punctuation holded by a player or who can he/she play against. All this rules go into this data layer.

- Entity model -> Represent data close to the db, usually db tables if we use a relational db. This could be tied to the domain but it does not have to. For example, imagine we want to store game punctuations in another table to have a coherent db design. Punctuation may go into UserDomain but no into UserEntity, you want to use entity model to map data from the db and there would be one table for each.

- DTO -> represents nothing concrete, it's just data. We may use this type of model to send data between layers. For example in an API response, you may want to send more data that you have in a UserDomain model. In a UserDomain we have things like userName, email etc. but you may want to attach some extra information that do not belong to the pure UserDomain (timestamp, some metadata needed by the frontend etc)


Right now we have this concepts a little bit mixed up in the project and would be nice if we could review them in the future, we can talk about it in the next meeting

Reply:
Thank you for the explanation :heart:. I think now I understand it a bit better :smile:.

For me, it's a bit confusing because I don't understand why is the `avatar` parameter from this class and the `avatar`, `id`, and `created_at` from the `UserEntity` class optional. I mean in TypeScript. Could we make them required?

Also, I don't understand why the `update()` and `add()` methods of the `UserRepository` take a `User` as a parameter and return a `User`.

I thought they could take a `Partial<UserEntity>` and return a `UserEntity` :confused:.

>I think right now we have these concepts a little bit mixed up in the project and would be nice if we could review them in the future, we can talk about it in the next meeting

Thank you, and I agree 100%!

Reply:
> For me, it's a bit confusing because I don't understand why is the avatar parameter from this class and the avatar, id, and created_at from the UserEntity class optional. I mean in TypeScript. Could we make them required?

Yep, you are right about this. They should be required because entity must map db object model. This optional thing in `UserEntity` is probably a left over. I'll amend this point as pointend in -> https://github.com/42AGV/ft_transcendence/pull/51#discussion_r895947568

Reply:
> Also, I don't understand why the update() and add() methods of the UserRepository take a User as a parameter and return a User.
> 
> I thought they could take a Partial<UserEntity> and return a UserEntity confused.

Mmm I see your point, and you may be (partially :smile:) right. The `UserRepository` is just an interface, the idea is the services communicate with it without knowing what db implementation is under the hood. And the `UserPostgresRepository` is the implementation for Postgres technology of that interface.

`UserEntity` is tied to the db and, therefore, to our Postgres implementation. We could use MongoDB or MySQL instead, that would mean our `UserEntity` would be slightly different. But we have to receive always the same (a User) from the repository no matter what specific implementation we have.

I know you may be thinking that User and UserEntity is almost the same, but not really. User can hold business logic since it belongs to the domain, whereas UserEntity is just data mapped to the db model.

There could also be additional problems if our models were very different (some fields in `UserEntity` missing in `User`, for example) In this case we should use additional [mappers](https://khalilstemmler.com/articles/typescript-domain-driven-design/repository-dto-mapper/). Since we don't need them right now, I've avoided them to not over engineer the code.

All this being said, it's true that there is something wrong here... https://github.com/42AGV/ft_transcendence/blob/da65933c93a622e677ab022b9a132abbb4fbdc59/transcendence-app/src/user/infrastructure/db/postgres/userPostgres.repository.ts#L20
All this specific methods should be returning `UserEntity` instead of `User` since they are Postgres implementations of the repo. I'll fix that :smile:

Reply:
> holded -> held
> Punctuation -> score

punctuation are these thingies-> , ; . -

> Entity model -> Represent data close to the db, usually db tables if we use a relational db. This could be tied to the domain but it does not have to. For example, imagine we want to store game punctuations in another table to have a coherent db design. Punctuation may go into UserDomain but no into UserEntity, you want to use entity model to map data from the db and there would be one table for each.

So if you want to have scores information, rather than being a field in the tables you would have Score being a separate table with... what kind of fields? Users? Like have a list of all users that have one point, all users that have 33 points, and so on? I fail to see how could such a mapping be relevant. Sorry, I don't get it.

> DTO -> represents nothing concrete, it's just data. We may use this type of model to send data between layers. For example in an API response, you may want to send more data that you have in a UserDomain model. In a UserDomain we have things like userName, email etc. but you may want to attach some extra information that do not belong to the pure UserDomain (timestamp, some metadata needed by the frontend etc)

This is confusing. So we have all data validation in UserDto right now. Is this wrong? Is it wrong to validate the data as soon as you receive it and error out as soon as you find out?

Reply:
Maybe a brief (or deep) explanation during this weekend meeting would be fine